### PR TITLE
Draft: Test that `/messages` works on remote homeserver and can be backfilled properly after many batches (MSC2716)

### PR DESCRIPTION
**Pull request by MadLittleMods**
Test that `/messages` works on remote homeserver and can be backfilled properly after many batches

Part of MSC2716: https://github.com/matrix-org/matrix-doc/pull/2716

Related to https://github.com/matrix-org/complement/pull/212

Used while developing https://github.com/matrix-org/synapse/pull/11114

### Dev notes

```go
	logrus.WithFields(logrus.Fields{
		"stack": string(debug.Stack()),
	}).Error("destroy stack")
```

```go
var firstEvent gjson.Result
var lastEvent gjson.Result
if len(eventsJson) > 0 {
	firstEvent = eventsJson[0]
	lastEvent = eventsJson[len(eventsJson)-1]
}
logrus.WithFields(logrus.Fields{
	"events":     len(eventsJson),
	"firstEvent": fmt.Sprintf("%s (%s)", firstEvent.Get("event_id"), firstEvent.Get("type")),
	"lastEvent":  fmt.Sprintf("%s (%s)", lastEvent.Get("event_id"), lastEvent.Get("type")),
	"start":      gjson.GetBytes(messsageResBody, "start").Str,
	"end":        gjson.GetBytes(messsageResBody, "end").Str,
}).Error("Number of events from `/messages`")
```

 - https://dave.cheney.net/2020/03/10/go-test-v-streaming-output
 - https://github.com/golang/go/issues/46959#issuecomment-947881974
 - https://github.com/golang/go/issues/27826

### CODE REVIEW

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:302]: Need to add marker event for a federated server to guarantee discovering the history

File: internal/client/client.go
Comment (MadLittleMods) [inline internal/client/client.go:313]: This adds the query parameters to the logged URL

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:1168]: Just making it easier to spot which messages were created at what point in the test and have some context when looking at an event ID.

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:1019]: This timeout is currently very large to accommodate the long ~20s `/messages` durations. We also have to make that request 11 times during the test which adds up very fast.

Synapse really has to chug for those requests 👹 and ideally wouldn't have to modify this at all.

I would need to look into optimizing Synapse to make this fast which we should probably do anyway as this is painfully slow.

File: internal/client/client.go
Comment (kegsay) [reply internal/client/client.go:313]: Thanks! :D

File: tests/msc2716_test.go
Comment (kegsay) [inline tests/msc2716_test.go:274]: Lovely comment but then you seem to use 2. Why?

File: tests/msc2716_test.go
Comment (kegsay) [inline tests/msc2716_test.go:258]: I don't think 2 batches is "many". The batch itself has many events though, so perhaps rename this test?

File: tests/msc2716_test.go
Comment (kegsay) [inline tests/msc2716_test.go:307]:
```suggestion
				// Grab the base insertion event ID to reference later in the marker event
```

File: tests/msc2716_test.go
Comment (kegsay) [inline tests/msc2716_test.go:314]: So this is new -> old?

File: tests/msc2716_test.go
Comment (kegsay) [reply tests/msc2716_test.go:1023]: This is currently set to a whopping 1000s by default https://github.com/matrix-org/complement/blob/136fd601570c43fbfb1cb10647884dff2338f4b4/internal/docker/deployment.go#L57 which seems excessive even given this.

File: tests/msc2716_test.go
Comment (kegsay) [inline tests/msc2716_test.go:1072]: Uhh, why? It's a test server with only us using it? 500ms waits seems excessive. 10ms seems okay if you are concerned logic errors would consume all CPU.

File: tests/msc2716_test.go
Comment (kegsay) [inline tests/msc2716_test.go:1062]: So this function doesn't actually check the events are in the correct order, just that they exist?

File: tests/msc2716_test.go
Comment (kegsay) [inline tests/msc2716_test.go:1044]: Why not use `JSONArrayEach`? https://github.com/matrix-org/complement/blob/136fd601570c43fbfb1cb10647884dff2338f4b4/internal/match/json.go#L164 It does these checks for you.

**Review by kegsay — APPROVED** [review]

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:1023]: @kegsay Sorry this wasn't clear as undrafting indicates "marked this pull request as ready for review" but I didn't assign you this one yet specifically because of this problem.

The test is good to go and `numBatches` could be switched back to `11` and shipped but want to make it actually acceptable time-wise to run against Synapse before merging.

I used `numBatches=2` during testing because it's much faster to see results while developing.

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:274]: See:

> I used `numBatches=2` during testing because it's much faster to see results while developing.
>
> https://github.com/matrix-org/complement/pull/214#discussion_r746829737

Running the tests against Synapse is extremely slow atm, https://github.com/matrix-org/complement/pull/214#discussion_r741678240

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:258]: It will be 11 when we ship, see https://github.com/matrix-org/complement/pull/214#discussion_r747128460

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:1072]: Removed from both places this was happening ⏩

---

This first spawned from the original MSC2716 Complement PR(https://github.com/matrix-org/complement/pull/68) for [`fetchUntilMessagesResponseHas`](https://github.com/matrix-org/complement/pull/68/files#diff-9868c2480f7570d7028a300cdf808ea451dee2ba46adc5a0bfd5fb0b5951a2eaR697). I think the thinking was that if the message response didn't have it, just give it some time to process everything and try again. But whatever, servers are fast 🤸‍♀️

Potentially lots of `/messages` requests is hard to debug if it takes just a bit of time to process all of the events. Just ran the tests without the delay and I don't see a big difference so should be fine 👍

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:1042]: I think it was because we can't break early with `JSONArrayEach` (can only break early if you throw an error which fails the test).

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:1060]: Correct, just a checkoff function

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:314]: The order doesn't matter because it's just a list of expected ID's, `expectedEventIDs`

We also just use it in check off function (`paginateUntilMessageCheckOff`) which doesn't care about the order.

---

There is a similar variable in another test where order does matter but we call it `expectedEventIDOrder`

File: tests/msc2716_test.go
Comment (kegsay) [reply tests/msc2716_test.go:1157]: Is the early return really a deal breaker? I don't think it'll affect runtime performance that much?

File: internal/client/client.go
Comment (MadLittleMods) [reply internal/client/client.go:313]: Split out to https://github.com/matrix-org/complement/pull/250

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:224]: TODO: Move this test under the `Federation` section once https://github.com/matrix-org/complement/pull/249 merges

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go]: Updated to re-use `JSONArrayEach`. It's not so bad ⏩

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:1175]: 👆 Calling out a change since the last review pass. I added a `denyListEventIDs` so we can ensure that the historical state doesn't appear between batches of historical messages while we're paginating.

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:99]: Reviewers note: A lot of the diff has been split out into separate PRs:

 - https://github.com/matrix-org/complement/pull/250
 - https://github.com/matrix-org/complement/pull/248
 - https://github.com/matrix-org/complement/pull/251
 - https://github.com/matrix-org/complement/pull/249

Comment (MadLittleMods) [reply tests/msc2716_test.go]: All of these are merged in now 🚀

**Comment by kegsay** [comment]
@MadLittleMods what is the status of this PR? It's still marked as a Draft.

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:1162]: > @MadLittleMods what is the status of this PR? It's still marked as a Draft.
>
> https://github.com/matrix-org/complement/pull/214#issuecomment-1123426080

Status is still the same as the last update in this thread. It's too slow on Synapse for me to be comfortable merging it yet so runs for everyone.

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:880]: Make it obvious that these batches are connected together

```suggestion
			t.Run("Backfill still works after many contiguous batches are imported", func(t *testing.T) {
```

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go:1204]: In terms of optimizing Synapse to make this test viable to run time-wise, I'm a bit blocked on a race condition in some recent code, https://github.com/matrix-org/synapse/pull/12394#discussion_r872012615

Comment (richvdh) [reply tests/msc2716_test.go:1204]: > In terms of optimizing Synapse to make this test viable to run time-wise, I'm a bit blocked on a race condition in some recent code, [matrix-org/synapse#12394 (comment)](https://github.com/matrix-org/synapse/pull/12394#discussion_r872012615) -> [matrix-org/synapse#12646](https://github.com/matrix-org/synapse/issues/12646)

Now https://github.com/matrix-org/synapse/issues/12988 I think

**Comment by kegsay** [comment]
Marked as stale; this PR will be closed in 7 days if there isn't any further progress. Please re-open when you get time. Long-lived branches are fine, but long-lived PRs like this always being in `Draft` is not fine.

File: tests/msc2716_test.go
Comment (MadLittleMods) [reply tests/msc2716_test.go]: Made progress on optimizing Synapse:

 - Saves 10s per `/messages` call: https://github.com/matrix-org/synapse/pull/13863 
 - Still need to figure out https://github.com/matrix-org/synapse/pull/13864

File: tests/msc2716_test.go
Comment (MadLittleMods) [inline tests/msc2716_test.go:1374]: Split out to https://github.com/matrix-org/complement/pull/489

**Pull request closed without merging**
